Migration to ESRP Code Signing from PackageES Code Signing

jefgen opened the pull request:
## Summary
This change modifies the MS-ICU Nuget build pipeline to use the ESRP Code Signing instead of the PackageES Code Signing.

## PR Checklist
* [x] I have verified that my change is specific to this fork and cannot be made upstream.
* [ ] I am making a maintenance related change.
* [ ] I am making a change that is related to usage internal to Microsoft.
* [x] I am making a change that is related to the Windows OS build of ICU.
* [ ] CLA signed. If not, please see [here](https://cla.opensource.microsoft.com/microsoft/icu) to sign the CLA.

## Detailed Description

Finally, after various delays and setbacks, we are now able to migrate to using ESRP Code Signing for the MS-ICU Nuget. This PR changes the Nuget build pipeline to use the ESRP Code Signing instead of the deprecated PackageES Code Signing. The end result in terms of the output signed binaries and packages is the same, but the mechanism is now different.

I did a test build of the pipeline and doubled-checked the output.
The DLLs are signed with the same certificate: `CN = Microsoft Code Signing PCA 2011`

The Nuget packages are signed with the same certificate as well:
```
Subject Name: CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
Issued by: CN=DigiCert SHA2 Assured ID Code Signing CA, OU=www.digicert.com, O=DigiCert Inc, C=US
```

For reference, here is the new build, using the ESRP Code Signing:
https://mscodehub.visualstudio.com/OSS_github.com_Unicode-org/_build/results?buildId=180999&view=results

Here is a previous build, using PackageES Code Signing (for comparison):
https://mscodehub.visualstudio.com/OSS_github.com_Unicode-org/_build/results?buildId=168347&view=artifacts&pathAsName=false&type=publishedArtifacts

The ESRP process signs the DLLs and packages *in-place*, rather than outputting them to a separate folder, so I also had to slightly modify the Nuget build script in order to not look for the "signed" folder anymore.

Note: This change also migrates us off the PackageES lab and uses the public Azure Pipelines for all stages now.

jefgen commented:
Note: After this PR is complete, I plan to bump the version number and update the changelog.

axelandrejs commented on build/azure-nuget.yml:
```diff
@@ -248,10 +248,15 @@ stages:
-    
+    variables:
+      signingPattern: |
+        icudt*.dll
```
Do we plan on ever signing icu.dll here? I guess not for now?

jefgen replied:
The icu.dll currently only builds/exists in the OS repo in Razzle. We'd first need to do work to port or "undock" it into this repo and create a Visual Studio project for it. So not adding here for now. :)

daniel-ju commented:
Do we need to permit the pipeline to access the code signing service every run?

jefgen commented:
> Do we need to permit the pipeline to access the code signing service every run?

When I setup the code signing service, I disallowed it access to all pipelines, except the manually triggered one. So the only pipeline that can use it is the Release-Nuget pipeline.

daniel-ju commented:
> When I setup the code signing service, I disallowed it access to all pipelines, except the manually triggered one. So the only pipeline that can use it is the Release-Nuget pipeline.

Does that mean the CI-Nuget on pull requests will never complete? It shows as waiting for permission before continuing with the pipeline, since it's trying to use the code signing service.

huichen123 commented on build/azure-nuget.yml:
```diff
@@ -302,15 +307,48 @@ stages:
+                }
+              ],
+              "toolName": "signtool.exe",
+              "toolVersion": "6.2.9304.0"
```
>"toolVersion": "6.2.9304.0" [](start = 14, length = 27)

what if the tool version is upgraded?

jefgen replied:
Hm, good question. I used the example from the ESRP code signing, but it looks like in ProjectReunion they don't use the `toolVersion` at all, so it must work without it.
I'll remove it in the next update. Thanks Hui. :)

huichen123 commented:
>     - task: PkgESSetupBuild@10

do you still need those pkges tasks?

---
Refers to: build/azure-nuget.yml:274 in 95182f9. [](commit_id = 95182f950e797398d94491736805d4fa68b273af, deletion_comment = False)

jefgen commented:
> > When I setup the code signing service, I disallowed it access to all pipelines, except the manually triggered one. So the only pipeline that can use it is the Release-Nuget pipeline.
> 
> Does that mean the CI-Nuget on pull requests will never complete? It shows as waiting for permission before continuing with the pipeline, since it's trying to use the code signing service.

Hmm, good point. It looks like it's stuck waiting for permissions. However, I really don't want to allow the "CI-Nuget" pipeline access to the signing service though.
Instead, I think I'm going to refactor the build pipeline so that the creation of the Nuget and the Signing of the Nguet are done as two separate steps. Then the signing stage can be entirely skipped in the `CI-Nuget` pipeline, but not in the `Release-Nuget` pipeline. (This is how the binaries code signing step works, and it doesn't block the pipeline currently).

I'll test this out in another branch, and then update this PR once I've been able to test things.

jefgen commented:
> do you still need those pkges tasks?

Good point. Based on looking at the ProjectReunion build infra, I don't think they are needed anymore.
I'll remove them in another update. Thanks!

jefgen commented:
**Update:** I tested it out, and it looks like the PackageES tasks can be removed without issue. I also removed the `toolVersion` from the sign task as well. (Thanks Hui!)

I've refactored the pipeline to move the Nuget signing into a separate stage from the Nuget package creation. This lets us completely skip the Nuget package signing for the `CI-Nuget` pipeline, but not in the `Release-Nuget` pipeline.

This means that we can enable the code signing service for _only_ the `Release-Nuget` pipeline, which must be manually triggered. Whereas the automatic CI pipeline doesn't have access to the code signing service connection at all.
(Thanks Daniel for noticing the CI pipeline was stuck.)

huichen123 commented on build/azure-nuget.yml:
```diff
@@ -248,10 +248,15 @@ stages:
     name: Package ES CodeHub Lab E
```
>Package ES CodeHub Lab E [](start = 10, length = 24)

there is no need to use PackageES lab, which has few build lines.

jefgen replied:
Can you code sign on non-PackageES lab agents?
FWIW, it looked like ProjectReunion was still using PackageES lab for the code signing part of the pipeline, which is why I kept this.

jefgen replied:
I'll try using a non-PackageES lab and see if it works....

jefgen replied:
Hm, I tried using the regular Azure Pipelines pool [here](https://mscodehub.visualstudio.com/OSS_github.com_Unicode-org/_build/results?buildId=181206&view=logs&j=27c746b8-fb0c-51f0-4fec-1863a64109c5). 

However, the CodeSign stage of the build appears to be cancelled with the following error message:
`##[error]The remote provider was unable to process the request.`

I wonder if maybe it doesn't support the ESRP task? Perhaps that is why ProjectReunion still uses PackageES lab?
Anyways, I think we might need to keep it using PackageES lab for now.

jefgen replied:
I tried a few more builds, and they had the same odd error, so leaving this as PackageES for now.

huichen123 replied:
not sure whether "Azure Pipelines" is a valid name. Reunion uses "pool:   vmImage: 'windows-2019'". 

---
In reply to: [612874376](https://github.com/microsoft/icu/pull/93#discussion_r612874376) [](ancestors = 612874376)

jefgen replied:
Hmm, interesting...

In the builds that failed I just used `pool: name: Azure Pipelines` as that is what I have setup for the initial build step here:
https://github.com/microsoft/icu/blob/master/build/azure-nuget.yml#L19
However, it also sets the `vmImage` as well. In the builds that failed, I did not set `vmImage` as well.

I wonder if both are perhaps needed?

I'll try out another build with both and see if that is it. Thanks Hui.

jefgen replied:
Ah hah! That was it! You need to set *both* the pool name, and the vmImage.
By setting, the code signing now works on Azure Pipelines!
This means we can completely move off the PackageES lab entirely. 

Thanks Hui for commenting on this and making me try out all the combinations. :)

huichen123 replied:
nice! btw, name is not required.

huichen123 commented on build/azure-nuget.yml:
```diff
@@ -434,29 +457,102 @@ stages:
+  - Build
+  - CodeSignBinaries
+  - CreateNuget
+  condition: and(eq(variables.codeSign, true), and(in(dependencies.Build.result, 'Succeeded'), in(dependencies.CodeSignBinaries.result, 'Succeeded', 'Skipped')))
```
>'Skipped' [](start = 149, length = 9)

what does this mean? Signed Nuget package with unsigned binaries inside?

jefgen replied:
Since it depends on `variables.codeSign` it will only really work for signed Nuget package. But good point, this extra condition is confusing. I'll remove it. 
Thanks!

jefgen commented:
**Update 2:** 
Note: This change also migrates us off the PackageES lab, and uses the public Azure Pipelines for all stages now.

huichen123 commented:
> # We need to set the version for the PackageES builds.

can you update the comment here?

---
Refers to: build/azure-nuget.yml:6 in 34a79d0. [](commit_id = 34a79d02deb990a5f9010ac22d480caa1d84c241, deletion_comment = False)

huichen123 commented:
> # Code signing only runs on the MSCodeHub build pipeline.

and here

---
Refers to: build/azure-nuget.yml:9 in 34a79d0. [](commit_id = 34a79d02deb990a5f9010ac22d480caa1d84c241, deletion_comment = False)

huichen123 reviewed (approved):
:shipit:

jefgen commented:
> >  We need to set the version for the PackageES builds.
> 
> can you update the comment here?

Good point, we can likely remove the `name` here.

> >  Code signing only runs on the MSCodeHub build pipeline.
> 
> and here
> 

Well the code signing still only runs on the pipeline in MSCodeHub, so this comment is still correct. However, I can update it a bit to note that the pipeline is manually triggered.
Thanks!
